### PR TITLE
NF: Improve listItemsAndMessage

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
Its inputs seemed stranged to me. A list of items, and an action that is decided according to the position. It seems to easy to get both value inconsistent.

I prefer the parameter to be a list of data, the data containing both the string and the onClick.

This ensure that each time the ListItem is created, it can easily be seen the text and the callback are consistent.

### CODE REVIEW

**Review by david-allison — COMMENTED** [review]
I'm fairly ambivalent, code wasn't good before, but I'm not sure if this moves the needle

```kotlin
                listItemData.map { data ->
                    ListItem(context.getString(data.stringRes)) {
                            dialog ->
                        data.onClick(context)
                        if (data.dismissesDialog) {
                            dialog.dismiss()
                        }
                    }
```

strikes me as more confusing due to the use of closures (regarding `data`).

**Review by david-allison — CHANGES_REQUESTED** [review]
Requesting changes (or rejecting if you don't feel it's worth the effort) based on further thought: 

There's little point in a refactoring which isn't associated with a functional change, and doesn't get the code into a 'good' state

File: AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/AndroidPermanentlyRevokedPermissionsDialog.kt [hunk 1]
```diff
@@ -54,13 +54,18 @@ object AndroidPermanentlyRevokedPermissionsDialog {
...
+                    ListItem(context.getString(data.stringRes)) {
+                            dialog ->
+                        data.onClick(context)
+                        if (data.dismissesDialog) {
+                            dialog.dismiss()
+                        }
+                    }
                 }
-            }
+            )
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/AndroidPermanentlyRevokedPermissionsDialog.kt:68]: Ah... GitHub lost the comment

A 'good' version of the code would look something like the following, and we should aim for this.

```kotlin
            listItemsAndMessage(
                message = message,
                listItems = StoragePermanentlyRevokedOptions.createList()
            )
```

**Review by lukstbit — CHANGES_REQUESTED** [review]
> Its inputs seemed stranged to me. A list of items, and an action that is decided according to the position. 

All AlertDialog methods that display a list of items use the pattern _set*Items(items, listener)_ (_listItemsAndMessage_ should have been called _setItemsAndMessage_), I don't see why we would move from this established pattern for this method only. Other reasons why I don't think this is an improvement:

- usually, all items in a list have the same action done on them so it would make sense to have a general listener instead of  an individual listener set at item level, an api should be built to accommodate the general use case
- seems more verbose with the extra class, also looking at DatabaseErrorDialog where we had just two items I don't want to think how building a list with 6-7 items would look like
- current code doesn't take in consideration the situation where I don't want to set a listener for item selection(for example there's a UI element in the list item that triggers an action)
- as a personal opinion, the current method reads better with the way we create the alert dialogs because everything is together(vs. having a _.map {ListItem(...)}_ somewhere outside of the _AlertDialog.Builder.create {...}_ lambda)

**Pull request closed without merging**